contrib/seekable_format/zstdseek_compress.c

```diff
@@ -0,0 +1,298 @@
+        return (size_t)zcs->maxFrameSize;
+    }
+
+    return (size_t)(zcs->maxFrameSize - zcs->frameDSize);
+}
+
+static size_t ZSTD_seekable_seekTableSize(ZSTD_seekable_CStream* zcs)
+{
+    size_t const sizePerFrame = 8 + (zcs->checksumFlag?4:0);
+    size_t const seekTableLen = ZSTD_skippableHeaderSize +
+                                sizePerFrame * zcs->framelog.size +
+                                ZSTD_seekTableFooterSize;
+
+    return seekTableLen;
+}
+
+static size_t ZSTD_seekable_writeSeekTable(ZSTD_seekable_CStream* zcs, ZSTD_outBuffer* output)
+{
+    BYTE* op = (BYTE*) output->dst;
+    BYTE tmp[4]; /* so that we can work with buffers too small to write a whole word to */
+
+    /* repurpose
+     * zcs->frameDSize: the current index in the table and
+     * zcs->frameCSize: the amount of the table written so far
+     *
+     * This function is written this way so that if it has to return early
+     * because of a small buffer, it can keep going where it left off.
+     */
+
+    size_t const sizePerFrame = 8 + (zcs->checksumFlag?4:0);
+    size_t const seekTableLen = ZSTD_seekable_seekTableSize(zcs);
+
+#define st_write32(x, o)                                                       \
+    do {                                                                       \
+        if (zcs->frameCSize < (o) + 4) {                                       \
+            size_t const lenWrite = MIN(output->size - output->pos,            \
+                                        (o) + 4 - zcs->frameCSize);            \
+            MEM_writeLE32(tmp, (x));                                           \
+            memcpy(op + output->pos, tmp + (zcs->frameCSize - (o)), lenWrite); \
+            zcs->frameCSize += lenWrite;                                       \
+            output->pos += lenWrite;                                           \
+            if (lenWrite < 4) return seekTableLen - zcs->frameCSize;           \
+        }                                                                      \
+    } while (0)
+
+    st_write32(ZSTD_MAGIC_SKIPPABLE_START | 0xE, 0);
+    st_write32(seekTableLen - ZSTD_skippableHeaderSize, 4);
+
+    while (zcs->frameDSize < zcs->framelog.size) {
+        st_write32(zcs->framelog.entries[zcs->frameDSize].cSize,
+                   ZSTD_skippableHeaderSize + sizePerFrame * zcs->frameDSize);
+        st_write32(zcs->framelog.entries[zcs->frameDSize].dSize,
+                   ZSTD_skippableHeaderSize + sizePerFrame * zcs->frameDSize + 4);
+        if (zcs->checksumFlag) {
+            st_write32(zcs->framelog.entries[zcs->frameDSize].checksum,
+                       ZSTD_skippableHeaderSize + sizePerFrame * zcs->frameDSize + 8);
+        }
+
+        zcs->frameDSize++;
+    }
+
+    st_write32(zcs->framelog.size, seekTableLen - ZSTD_seekTableFooterSize);
+
+    if (output->size - output->pos < 1) return seekTableLen - zcs->frameCSize;
+    if (zcs->frameCSize < seekTableLen - 4) {
```

Comment:
It's the same logic that's present in `st_write32` to avoid writing data we've written in a previous call.  Without it, the seekable frame descriptor could be written multiple times, instead of writing the magic number.